Added whitelisting of procedures

eebus opened the pull request:
changelog: Added support for whitelisting procedures. This allows a user to load only a specific procedures from a package of procedures. The setting is to use is  `dbms.security.procedures.white_list`

eebus committed:
Added white listing of procedures

eebus committed:
Improved the config description and name. Also some cleanup

eebus committed:
Changed default to * instead of empty.

eebus committed:
Improved descriptions.

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureCompiler.java:
```diff
@@ -175,7 +175,18 @@
+
+                if ( !(fullAccess || config.whiteListed( procName.toString() )) )
+                {
+                    log.warn( String.format( "The procedure '%s' is not white listed.", procName.toString() ) );
```
whitelisted is one word.
Perhaps consider something along the lines of `The procedure '%s' is not on the whitelist and won't be loaded`?

OliviaYtterbrink commented on community/kernel/src/test/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureTest.java:
```diff
@@ -305,6 +309,65 @@ public void shouldSupportProcedureDeprecation() throws Throwable
+    }
+
+    @Test
+    public void shouldLoadFullAccessProcedure() throws Throwable
```
This doesn't seem to test what it says, and looks very similar to `shouldLoadWhiteListedProcedure `

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -502,10 +502,15 @@
+    @Description( "A list of procedures (comma separated) that are to be loaded." +
+            " If this setting is left empty all procedures will be loaded." )
+    public static final Setting<String> procedure_white_list =
+            setting( "dbms.security.procedures.white_list", Settings.STRING, "*" );
```
white_list should be whitelist

OliviaYtterbrink replied:
Perhaps add something about it supporting wildcards

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ReflectiveProcedureCompiler.java:
```diff
@@ -175,7 +175,18 @@
+                String definedName = method.getAnnotation( Procedure.class ).name();
+                QualifiedName procName = extractName( procDefinition, method, valueName, definedName );
+
+                if ( !(fullAccess || config.whiteListed( procName.toString() )) )
```
Maybe switch order of the if and else part of this?
`fullAccess || config.whiteListed( procName.toString()` is easier to read and understand

OliviaYtterbrink commented on community/kernel/src/main/java/org/neo4j/kernel/impl/proc/ProcedureConfig.java:
```diff
@@ -104,6 +109,11 @@ boolean fullAccessFor( String procedureName )
         return accessPatterns.stream().anyMatch( pattern -> pattern.matcher( procedureName ).matches() );
     }
 
+    boolean whiteListed( String procedureName )
```
`whitelisted` or maybe `isWhitelisted`?

OliviaYtterbrink reviewed (changes requested):
Looks good for the most part, just a few changes needed.

eebus committed:
Improved naming and testing.
Also now supports only wildcard character '*' and not all of regex in the procedure lists.

eebus committed:
Fixed style check
